SEP-0012.md

```diff
@@ -0,0 +1,495 @@
+
+        When a slice is applied to a NDCube object it must be applied to the following attributes:
+
+          * data
+          * wcs
+          * uncertainty
+          * mask
+          * extra_coords
+
+        the following attributes must be copied unaltered to the returned NDCube:
+
+          * unit
+          * meta
+
+        """
+
+    @abc.abstractmethod
+    @property
+    def extra_coords(self) -> ExtraCoordsABC:
+        """
+        Coordinates not decribed by ``NDCubeABC.wcs`` which vary along one or more axes.
+        """
+
+    @abc.abstractmethod
+    @property
+    def global_coords(self) -> GlobalCoordsABC:
+        """
+        Coordinate metadata which applies to the whole cube.
+        """
+
+    @abc.abstractmethod
+    @property
+    def wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube.
+
+        If a `BaseLowLevelWCS` is provided to the constructor it should be
+        "upgraded" to a high level object.
+        """
+
+    @abc.abstractmethod
+    @property
+    def combined_wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube, including the coordinates specified in ``.extra_coords``.
+
+        This transform should implement the high level wcsapi, and have
+        `pixel_n_dim` equal to the number of array dimensions in the
+        `.NDCube`. The number of world dimensions should be equal to the
+        number of world dimensions in ``self.wcs`` and in ``self.extra_coords`` combined.
+        """
+
+    @abc.abstractmethod
+    @property
+    def array_axis_physical_types(self) -> Iterable[Tuple[str, ...]]:
+        """
+        Returns the WCS physical types associated with each array axis.
+
+        Returns an iterable of tuples where each tuple corresponds to an array axis and
+        holds strings denoting the WCS physical types associated with that array axis.
+        Since multiple physical types can be associated with one array axis, tuples can
+        be of different lengths. Likewise, as a single physical type can correspond to
+        multiple array axes, the same physical type string can appear in multiple tuples.
+        """
+
+    @abc.abstractmethod
+    def axis_world_coords(self,
+                          *axes: Union[int, str],
+                          edges: bool = False,
+                          wcs: Optional[Union[BaseHighLevelWCS, ExtraCoordsABC]] = None
+                          ) -> Iterable[Any]:
+        """
+        Returns world coordinate values of all pixels for all axes.
+
+        Parameters
+        ----------
+        axes
+            Axis number in numpy ordering or unique substring of
+            `.NDCube.wcs.world_axis_physical_types` or
+            `.NDCube.wcs.world_axis_names` of axes for which real world
+            coordinates are desired. ``axes=None`` implies all axes will be
+            returned.
+
+        edges
```

Comment:
This needs a definition of what an `edge` is. Is it a 1D edge, or is a (N-1)D surface, e.g. in 3D it would be the 6 2D surfaces on the outside of a cube.

Reply:
Think of pixels like histogram bins.  If ``edges=False`` (default) coordinate values at the pixel/bin centers is returned.  If ``edges=True`` the coordinate values at the pixel/bin edges will be returned.  So the length of the returned coordinate objects will have a length in each dimension 1 greater than the number of array elements.

The equation/pseudo-code way of thinking of it is if ``edges=False``, the values entered to ``WCS.pixel_to_world`` is ``np.arange(0, n)``.  If ``edges=True`` the pixel values entered are ``np.arange(-0.5, n+0.5)``

Reply:
I would rename this argument `corners` instead of `edges`. In > 1D, `edges` is ambiguous - are the points on one edge halfway along the pixels in the other dimension, or at the corners?  `corners` is un-ambiguous in this respect.

Reply:
I like that suggestion. This being Danny's corner of the API I will let him comment before implementing that.

Reply:
To me, `corners` suggests the values at the lower edge of the first pixel and the upper edge of the last pixel in each array dimension.  So for a 2-D cube, no matter how many pixels you have, I would expect `corners` to return coords at only 4 pixel locations.  Whereas `edges=True` should return the real world coords at the edges of all pixels, as opposed to the default of pixel centers.

I suggest `pixel_edges`.

Reply:
How about `pixel_corners`? I still think `pixel_edges` is ambiguous, as an edge is a continuous set of points, whereas a corner is a well defined single point.

Reply:
Ah, I see your point when you talk about an edge as a continuous set of points.  This is true when ndim > 1 and the axes are dependent.  I was coming from the idea that axes are independent and so pixels can be separated into 1-D bins, like in a 1-D histogram.  In this case we talk about bin centers and bin edges.  The continuous edge collapses to a single point.

Having thought it through some more, I am OK with `pixel_corners` as it is probably slightly more generalised.  But I think we will have some users for whom this will be a bit less less intuitive.